Symex complexity heuristic

JohnDumbell opened the pull request:
This is a heuristic that tries to predict when we're entering an area that symex will start to get bogged down in, which will either result in extreme running times and/or an equation that's unrealistically solvable. When it detects this is happening it simply cuts off that branch. 

In practice this can be used as an intelligent depth limit or timeout in symex, forcing it to only consider paths below a certain resistance. This has proven very effective in codebases that for the most part are solvable but have particular core methods that are written in such a way that simply entering them can cause a run to go from minutes to hours or more. 

As each branch executes it keeps track of its own internal complexity. This is then used in two distinct ways:

1. When a branch' complexity rating is above a threshold it falses the guard and then marks that branch as failed.
2. When enough branches have failed, we work out if these branches are part of a loop somewhere up the stack. If we are, we take the lowest-executing loop and blacklist it (any branch entering it immediately gets cancelled) for all currently-running loops.

For example, with this code:
```
loopA:
    loopB:
        ... complex operations
    
    loopC: 
        ... simple operations
```
`loopB` will be blacklisted early, and then for the rest of the iterations of `loopA` it'll only run `loopC`. But as soon as this loop is finished that limitation will go, and when this code is run in a different context then it'll have a chance to execute again. Blacklists are not global. This is needed due to the way loops work, as even if you kill off every branch of one run it'll just try again * the unwind limit, made worse if there's recursion involved.

The base heuristic is uses is guard size, but it's written in a way to abstract what you pass into the command-line vs what limitations are enforced internally. As long as `--symex-complexity-limit 4` gives similar overall runtime guarantees, what is actually being used internally to work this out is irrelevant. It's likely that the heuristic _will_ need to change over time as things change internally to symex, but at least for now guard size is a direct mapping to the relative complexity of conditions that is being written into the equation, so can be used as a good, quick-to-compute measurement of how complex a certain expression is at any one time. 

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

JohnDumbell committed:
Add size() to exprt to count nodes

JohnDumbell committed:
Move symex_configt into its own file

JohnDumbell committed:
Add commandline options for symex_complexityt

JohnDumbell committed:
Initial commit of symex complexity work

This work allows for a more intelligent way of pathing  around / slicing
away problematic code dynamically as symex runs. Every path will have a
'complexity' rating that shows currently how difficult the code it's
running is for symex to analyze (alternatively also for the SAT solver
to solve) and when that rating breaches a certain range, that path will
be shut down.

This means we won't be able to generate traces that would use the
discarded path, but when that path is not important it's a net benefit,
and when it was important CBMC wouldn't have completed anyway so we
trade an inaccurate trace for nothing at all.

Note: There is special consideration for loops: when any of a loops
children fails (inc. loops / recursion) it increments a counter, and
when that counter reaches a certain limit the whole loop itself gets
cancelled. This is done because when a branch deep within a loop gets
cancelled along all paths the loop will simply try again even though
it's been proven that the code will become too complicated to analyze.

allredj commented on src/util/expr.h:
```diff
@@ -84,6 +84,19 @@ class exprt:public irept
   }
 
+  /// Amount of nodes this expression tree contains.
+  std::size_t size() const
```
Document that complexity is linear in the tree size, so that we are warned against using it frivolously.

allredj commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,126 @@
+#include "symex_complexity.h"
+#include "goto_symex_state.h"
+
+bool symex_complexityt::is_too_complex(goto_symex_statet &state)
```
Although `is_too_complex` is a good name for a >100 line function ;-), I guess it doesn't properly convey very well that its main effect is a side-effect.

smowton replied:
Agreed, perhaps name this `check_complexity`

owen-mc-diffblue commented on src/goto-symex/symex_config.h:
```diff
@@ -31,6 +31,16 @@ struct symex_configt final
 
+  /// Arbitrary number to represent how complex a branch of symex should get
+  /// before it's considered unviable for completion.
+  /// This has no relation to the underlying mechanism for what is copmlexity,
```
Typo in "copmlexity"

smowton replied:
"for what is complexity" -> "for computing complexity"?

smowton commented on src/goto-programs/goto_program.cpp:
```diff
@@ -532,10 +533,33 @@ std::string as_string(
+  {
+    bool loop_target = false;
+    for(auto &inc_edge : i.incoming_edges)
+      loop_target = loop_target || inc_edge->is_backwards_goto();
```
Check that number of incoming edges is at least 2 as well, otherwise we're looking at an ordinary `goto` instruction

smowton commented on src/goto-programs/goto_program.cpp:
```diff
@@ -532,10 +533,33 @@ std::string as_string(
+      i.loop_number = active_loops.back();
+    }
+
     if(i.is_backwards_goto())
```
This is already the case in goto-symex, but suggest commenting explicitly that this notion of loop is inadequate, backward-gotos are not necessarily in 1-to-1 correspondence with loops.

smowton commented on src/goto-programs/goto_program.cpp:
```diff
@@ -532,10 +533,33 @@ std::string as_string(
+      active_loops.push(nr++);
+    }
+
+    // Assign the loop number to every instruction associated with that loop.
```
Explicitly note this is the innermost loop

smowton commented on src/goto-symex/Makefile:
```diff
@@ -35,6 +35,7 @@ SRC = auto_objects.cpp \
       symex_target.cpp \
       symex_target_equation.cpp \
       symex_throw.cpp \
+      symex_complexity.cpp \
```
commit message for "Initial commit of symex complexity work": "This means that our trace will miss context" -> "This means we won't be able to generate traces that would use the discarded path"

smowton replied:
Also we don't generate partial traces because of this, we generate normal traces but may fail to spot some which otherwise would have been generated

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,126 @@
+  // Check if this branch is too complicated to continue.
+  if(complexity >= max_complexity)
+  {
+    log.warning() << "Branch considered too complex"
```
Note this message comes from symex

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,126 @@
+        in_loop = true;
+        if(loop_iter.second.blacklisted_loops.count(current_loop_id) != 0)
+        {
+          log.warning() << "Blacklisted loop operation attempted"
```
Again note this comes from symex

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -73,14 +73,47 @@ void symex_transition(
+    // And conversely, if we're transitioning out of a loop make sure
+    // we state that it's not active anymore.
+
+    // Match the do-while pattern.
```
Again note in comments that this is dubious and should be replaced with a proper loop analysis

JohnDumbell commented:
Finished off and tested the remaining aspects that weren't complete. Will start to have a look at proper loop detection as what I'm using for nested loops is very rough, let alone the code around finding out if a loop is active or not. 

Will also run this against some benchmarks to see how much complexity slows down execution when it doesn't cut off any branches and tweak if it is significantly detrimental.

smowton commented:
Note we have `analyses/natural_loops.h`. I'm currently working on improving how the dominator analysis it depends on works, with the goal that it should be fast enough that symex can dominator-analyse and loop-analyse every function it works on without that having any significant impact time-wise.

romainbrenguier commented on src/util/expr.h:
```diff
@@ -84,6 +84,19 @@ class exprt:public irept
 
+  /// Amount of nodes this expression tree contains.
+  std::size_t size() const
+  {
```
I think the definition should go to the cpp file, this should probably not be inlined

romainbrenguier commented on src/goto-symex/symex_config.h:
```diff
@@ -0,0 +1,53 @@
+
+Module: Symbolic Execution
+
+Author: Daniel Kroening, kroening@kroening.com
```
:question: should it be yourself or diffblue as an author?

JohnDumbell replied:
Since I was just moving it from another file with no changes just left it with the original files author - no real preference though!

smowton replied:
Change the module to "Symbolic Execution Config"; inheriting authorship is fine IMHO

romainbrenguier commented on src/goto-symex/symex_config.h:
```diff
@@ -0,0 +1,53 @@
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
```
I think you will need `\file` documentation

romainbrenguier commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -197,6 +197,15 @@ void jbmc_parse_optionst::get_command_line_options(optionst &options)
 
+  if(cmdline.isset("symex-complexity-limit"))
+    options.set_option(
+      "symex-complexity-limit", cmdline.get_value("symex-complexity-limit"));
```
use braces (`{ }`) around multi-line `if`s

romainbrenguier commented on src/goto-checker/bmc_util.h:
```diff
@@ -219,6 +221,13 @@ void run_property_decider(
   " --no-self-loops-to-assumptions\n" \
   "                              do not simplify while(1){} to assume(0)\n" \
   " --no-pretty-names            do not simplify identifiers\n" \
+  " --symex-complexity-limit     how complex a path can become before symex\n" \
```
an argument symbol should appear after the option, and its meaning should be documented: what's the measure of complexity used?

romainbrenguier commented on src/goto-programs/goto_program.cpp:
```diff
@@ -532,10 +533,36 @@ std::string as_string(
+    bool loop_target = false;
+    if(i.incoming_edges.size() >= 2)
+      for(auto &inc_edge : i.incoming_edges)
+        loop_target = loop_target || inc_edge->is_backwards_goto();
```
Use `std::any_of`. Then `loop_target` can be defined by a simple expression, marked const and maybe inlined below

romainbrenguier commented on src/goto-symex/frame.h:
```diff
@@ -42,8 +42,37 @@ struct framet
+    }
+
+  private:
+    int nesting_level = -1;
```
This is a bit unclear, I guess -1 is used too mean this has not been computed yet. To make that clearer you could replace this field and the `get_nesting_level()` method by a field `lazyt<int> nesting_level = lazy([]{return parent_loop == nullptr ? 0 : parent_loop->nesting_level.force() + 1;})`

romainbrenguier commented on src/goto-symex/goto_state.h:
```diff
@@ -63,6 +63,20 @@ class goto_statet
+  std::size_t recorded_complexity = 0;
+
+public:
+  void record_complexity()
```
It's not clear what is going on here. Is it meant to avoid recomputing guard size everytime? in that case use a `lazyt`, otherwise this probably shouldn't be part of this class

JohnDumbell replied:
It's cleaner to get the states to record their own complexity, and this method is just an abstraction as recording complexity can be anything, that it's so simple right now is just this particular implementation. Just saves time in the future for no real downside.

(The alternate approach is get the `complexity_module` to record everything, but that introduces more code complexity and a few other downsides that really isn't worth it)

romainbrenguier replied:
> It's cleaner to get the states to record their own complexity

I'm not sure about that, if objects record information about the data that is in them, then it gets confusing. Right now I have no idea about what this complexity represents and when it should be updated. Is there a relation between this complexity and the other fields which should be preserved? If so, the interface of the class should ensure it is preserved.

JohnDumbell replied:
Shrunk it to what's really required right now, while still keeping in line with how it's meant to be used. If multiple recording/state is needed later it can be trivially put in anyway.

romainbrenguier commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,196 @@
+Author: John Dumbell
+
+\*******************************************************************/
+
```
needs a `\file` documentation I think

JohnDumbell commented:
Taking away the WIP label as this is ready for active reviews. I'm not going to use our `natural_loops` implementation yet as it isn't fast enough right now for this particular usage, and it would only increase accuracy. It can be used later once it's got a little better.

I'll write one or two tests to assert that limitations are honoured, and will also act as a gauge when the heuristic needs to be changed or updated after people are relatively happy with the concept and code.

romainbrenguier commented on src/goto-symex/goto_state.h:
```diff
@@ -63,6 +63,13 @@ class goto_statet
 
+  /// Get the currently calculated complexity for this state. Used
+  /// to short-circuit states if they have become too computationally complex.
+  inline std::size_t get_complexity()
```
`get_` is misleading since you would expect it to return the value of a field. `complexity()` is fine

romainbrenguier commented on src/goto-symex/goto_state.h:
```diff
@@ -63,6 +63,13 @@ class goto_statet
+  /// to short-circuit states if they have become too computationally complex.
+  inline std::size_t get_complexity()
+  {
+    return guard.as_expr().size();
```
:bulb: this would be costly for BDDs. Ideally we should have a `complexity` method for guards, so that it can be substituted to something more efficient (you don't need to do it in this PR)

JohnDumbell replied:
Yeah, it was something I was musing on and was going to talk with you at some point or do a little bit more poking around to find a proper solution.

romainbrenguier replied:
I think the proper solution for BDDs would be to use something like the dagSize method of Cudd http://web.mit.edu/sage/export/tmp/y/usr/share/doc/polybori/cudd/cuddAllDet.html#Cudd_DagSize
but I'm not sure there is something similar for miniBDD

owen-mc-diffblue replied:
I would add a comment here saying that if we start using BDDs for guards then we should use something more efficient, like dagSize.

romainbrenguier commented on src/goto-symex/goto_state.h:
```diff
@@ -63,6 +63,13 @@ class goto_statet
   /// Threads
   unsigned atomic_section_id = 0;
 
+  /// Get the currently calculated complexity for this state. Used
```
"currently calculated" should be removed

romainbrenguier commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,197 @@
+        loop_iter.second.is_active &&
+        loop_iter.second.blacklisted_loops.count(loop_id) != 0)
+      {
+        return true;
```
looks like a `std::any_of`

romainbrenguier commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,197 @@
+    for(auto &loop_iter : frame_iter->loop_iterations)
+    {
+      if(loop_iter.second.is_active)
+        info_vectors.emplace_back(loop_iter.first, &loop_iter.second);
```
:pick: info_vectors could be made `const` if using `make_range(frame_iter->loop_iterations).filter(...).map(...)`

romainbrenguier commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,197 @@
+  // otherwise we will continue to execute nested loops that we already know
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+  irep_idt blacklist;
```
the meaning of this irep_idt is not clear

smowton replied:
loop_to_blacklist?

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+  symex_complexityt(message_handlert &logger, const optionst &options)
+    : log(logger)
+  {
+    std::size_t limit = options.get_signed_int_option("symex-complexity-limit");
```
The definition should go to the cpp file

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+
+      std::size_t failed_children_limit =
+        options.get_signed_int_option("symex-complexity-failed-children-limit");
+      std::size_t unwind = options.get_signed_int_option("unwind");
```
const

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+      // while larger numbers are very large to represent what you're asking.
+      max_complexity = static_cast<std::size_t>((limit * limit) * 75);
+
+      std::size_t failed_children_limit =
```
const

romainbrenguier commented on src/util/expr.h:
```diff
@@ -84,6 +84,9 @@ class exprt:public irept
   }
 
+  /// Amount of nodes this expression tree contains.
+  std::size_t size() const;
```
:bulb: This doesn't take into account sharing. Maybe it should, or we could have two versions of it.

JohnDumbell replied:
Sharing shouldn't matter unless a shared tree requires a different way of counting. Whether you have 5 expressions that are all shared or 5 separate, their size should all be considered the same.

smowton replied:
Perhaps just mention it in the docstring -- `x + x` has `2|x| + 1` nodes, regardless of whether the two expressions are shared or not (i.e. this is the size of the printed tree, not the memory consumption of the expression)

romainbrenguier commented on src/goto-symex/symex_heuristic_transformation.h:
```diff
@@ -0,0 +1,23 @@
+
+/// Default heuristic transformation that cancels branches when complexity
+/// has been breached.
+class symex_heuristic_transformationt
```
I'd prefer to make this class abstract, to show it is an interface, or even replace it by `std::function<void(const complexity_violationt, goto_symex_statet &)>`

romainbrenguier commented on src/goto-symex/symex_heuristic_transformation.h:
```diff
@@ -0,0 +1,23 @@
+public:
+  void transform(
+    const complexity_violationt heuristic_result,
+    goto_symex_statet &current_state)
```
should it be virtual?

romainbrenguier commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,197 @@
+      {
+        if(
+          !active_loop || active_loop->loop_info->nesting_level.force() <
+                            loop_iter.second.nesting_level.force())
```
it looks like you should use `std::max_element`

romainbrenguier commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,197 @@
+    // As we're walking bottom-up, the first frame we find with active loops
+    // is our closest active one.
+    if(active_loop)
+      return active_loop;
```
it looks like this loop should be replaced by a `std::find_if`

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+
+  /// Default heuristic transformation. Adds unconditional false guards to
+  /// state.
+  symex_heuristic_transformationt default_transformation;
```
Is there something else than the default? What is the procedure for choosing a different heuristic? Should this class be a template argument? Why do we have an instance of a class which has no field?

JohnDumbell replied:
1. No.
2. That's up for future people to decide. The entirely abstract design is this: you use [specified] heuristic that determines when a certain violation happens, which then runs a selection of [transformations] that can do many things: modify guards, modify the equation, modify all sorts of data structures, but all with the goal to ease the burden of whatever's happening. It's doesn't live up to that design right now, because I wanted to get things working before I went all-in, then I could decide from there how far I wanted to take it.
3. Don't think that's needed.
4. Is your point here that it's effectively a static method that doesn't really need an instance of anything, so why going through the trouble of creating an object at all? While that's correct, I prefer the idea that even the default has to adhere to the same API as all other transformations. This is a personal preference thing.

romainbrenguier replied:
I think this is alright, but it looks like you are trying to make the design be adaptable for a very generic class of heuristics, while there is no need for such customization level in the current implementation. This makes reasoning about the class currently more difficult than it would need to be, without being sure that it will not need to be changed in the future. 
I would prefer the design to start simple and close to what the first implementation needs (for instance make it a function instead of an abstract class) and then generalize as needed, (but you don't need to change it now).

JohnDumbell replied:
I prefer it being simple but also having the frameworks to support it in the future. It's the balance between building for right now and giving people hints and hooks to extend it in the future, while not wasting time building things that people will never use.

I don't mind simplifying it, as I was intending too anyway, but I'll still be keeping to that ideal either way.

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+  static optionalt<loop_complexity_infot>
+  get_current_active_loop(call_stackt &current_call_stack);
+
+public:
```
the public interface should go first

romainbrenguier commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,70 @@ void symex_transition(
+              !current_loop_info.parent_loop ||
+              current_loop_info.nesting_level.force() <
+                loop.second.nesting_level.force())
+              current_loop_info.parent_loop = &loop.second;
```
`std::max_element` ?

romainbrenguier commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,70 @@ void symex_transition(
+    }
+
+    // Otherwise do for-each / while.
+    for(const auto &i_e : state.source.pc->incoming_edges)
```
i_e --> incoming_edge

romainbrenguier commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,70 @@ void symex_transition(
+        if(loop_entry_iter != frame.loop_iterations.end())
+        {
+          loop_entry_iter->second.is_active = false;
+          break;
```
looks like this loop should be replaced by a `std::find_if`

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+#include "goto_symex_state.h"
+#include "symex_heuristic_transformation.h"
+
+class loop_complexity_infot
```
needs documentation and could be a struct since everything is public

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  irep_idt loop_id;
+  framet::loop_infot *loop_info;
```
Could we avoid a raw pointer here? Or at least document why it is safe to store it here.

JohnDumbell replied:
This is one of the main reasons I wanted someone with more C++ experience than me pitch in. Tried a few nicer options but couldn't work because of various reasons, so I'd like to know their preferred way of dealing with things like this.

Also the self-referencing pointer loop_infot too.

romainbrenguier replied:
I'm not exactly sure about what you need here, but seeing how it is used, I think you could use a frame iteraror + a loop_id and that should be enough to retrieve the info you need. The fact that the type is an iterator would make it clear that it shouldn't be used outside the scope of the container which was given as input to the function returning this.

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+  /// we're in a loop further up the stack that will still be returned as
+  /// the 'active' one.
+  static optionalt<loop_complexity_infot>
+  get_current_active_loop(call_stackt &current_call_stack);
```
can it really mutate the currect_call_stack? :duck:

JohnDumbell replied:
Yeah, it grabs a pointer and mutates it later.

romainbrenguier commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,155 @@
+  /// Checks whether we're in a loop that is currently considered blacklisted,
+  /// and shouldn't be executed.
+  static bool
+  in_blacklisted_loop(call_stackt &current_call_stack, const irep_idt &loop_id);
```
:duck:

JohnDumbell replied:
This one though can be const, changed it.

smowton commented on src/goto-programs/goto_program.cpp:
```diff
@@ -532,10 +533,36 @@ std::string as_string(
 void goto_programt::compute_loop_numbers()
 {
-  unsigned nr=0;
+  // This deals with nested loop number assignment by treating the queue
```
This needs some testing regarding what happens when abnormally shaped GOTO programs are used -- e.g. entry in the middle of a loop:
```
// Enter the loop:
goto entry;

head:
  ...
entry:
  ...
goto head;
```

Or a loop with multiple backedges (more than one `goto head;` statement), or a loop with multiple edges:

```
if(x)
  goto entry1;
else
  goto entry2;

entry1:
  if(iterate())
    goto entry2;

entry2:
  if(iterate())
    goto entry1;
```

We might be able to get by with a hackish loop characterisation or might need the real natural-loops pass, but either way we should have a go at breaking it and check something halfway sensible happens.

smowton commented on src/goto-programs/goto_program.cpp:
```diff
@@ -540,10 +540,12 @@ void goto_programt::compute_loop_numbers()
-        loop_target = loop_target || inc_edge->is_backwards_goto();
+    bool loop_target =
+      i.incoming_edges.size() >= 2 &&
+      std::any_of(
```
Merge this change into previous commit

smowton commented on src/goto-programs/goto_program.cpp:
```diff
@@ -540,10 +540,12 @@ void goto_programt::compute_loop_numbers()
-    if(i.incoming_edges.size() >= 2)
-      for(auto &inc_edge : i.incoming_edges)
-        loop_target = loop_target || inc_edge->is_backwards_goto();
+    bool loop_target =
```
const

smowton commented on src/goto-symex/complexity_violation.h:
```diff
@@ -0,0 +1,24 @@
+/// Loop: If we've violated a loop complexity boundary, or are part of
+/// a blacklisted loop.
+/// Branch: If this particular branch has been deemed too complex.
+enum complexity_violationt
```
use `enum class` (strong enum, not implicitly convertable to int / other enum, and uses `complexity_violationt::` as a name prefix rather than polluting the global namespace)

smowton commented on src/goto-symex/frame.h:
```diff
@@ -42,8 +42,35 @@ struct framet
+    int nesting_level = 0;
+
+    /// Sets parent and calculates current nesting level.
+    bool set_parent(loop_infot *par_loop)
```
par -> new_parent

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+  if((complexity_active = limit > 0))
+  {
+    // This gives a curve that allows low limits to be rightly restrictive,
+    // while larger numbers are very large to represent what you're asking.
```
what you're asking?

JohnDumbell replied:
Was a rather ambiguous reference to the idea that asking for complexity 6 is not as simple as stating that you're allowing double the complexity of 3. Each number is an individual statement of how much complexity you want in a vague curve shape and not directly related to the last (or future) beyond the idea that 1 = super strict, 20 = super lenient.

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+    if(failed_children_limit > 0)
+      max_loops_complexity = failed_children_limit;
+    else if(unwind > 0)
+      max_loops_complexity = std::max(static_cast<int>(floor(unwind / 3)), 1);
```
justify the /3 somehow?

JohnDumbell replied:
2/7, 1/3, 3/10 are all decent numbers for how many branches you'd allow to fail when using this system. Nothing more than that really - if you want more refined control I'd expect you to pass an exact number.

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+  }
+}
+
+/// Takes into account nested loops and various other sublties.
```
sp subtleties

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+
+  // If non-empty, is the lowest-level loop we should blacklist in the stack.
```
blacklisted?

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+bool symex_complexityt::are_loop_children_too_complicated(
+  call_stackt &current_call_stack)
+{
+  // Check that current loops don't violate complexity boundaries, if they
```
boundaries -> limits

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+
+    for(loop_complexity_infot &loop_comp : info_vectors)
+    {
+      // If we've already violated complexity, add to all parent loops
```
a complexity limit

smowton commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,229 @@
+      if(are_loop_children_too_complicated(current_call_stack))
+      {
+        log.warning()
+          << "[symex-complexity] Loop operations considered too complex"
```
Mention in the log message what we're doing, not just what we observed (e.g. "skipping", "dropping path", "assuming not entered / taken"

smowton commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,139 @@
+
+/// Symex complexity module.
+///
+/// Dynamically cancels branches when symex considers them too complex.
```
Define "cancels"

smowton commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,139 @@
+/// Symex complexity module.
+///
+/// Dynamically cancels branches when symex considers them too complex.
+/// A branch is too complex when runtime will be beyond reasonable
```
will -> may

smowton commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,139 @@
+  complexity_violationt check_complexity(goto_symex_statet &state);
+
+  /// Runs a suite of transformations on the state and symex executable,
+  /// performing whatever transformations are deemed right depending on
```
deemed right -> required

smowton commented on src/goto-symex/symex_complexity.h:
```diff
@@ -0,0 +1,139 @@
+  /// Checks the passed-in state to see if its become too complex for us to deal
+  /// with, and if so set its guard to false.
+  /// \param state goto_symex_statet you want to check the complexity of.
+  /// \return True/false depending on whether this branch has been abandoned.
```
has been -> should be? Or does this actually do the work itself (in which case rename accordingly, something like `truncate_path_if_too_complex`; at present the name suggests a const method

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,65 @@ void symex_transition(
+        {
+          if(
+            loop.second.is_active && !loop.second.is_recursion &&
+            current_loop_info.nesting_level < loop.second.nesting_level)
```
This reads backwards -- I'd expect this to mean "current_loop_info's parent is now loop.second", but this is surely the other way around given `loop.second` is *more* deeply nested

smowton replied:
Also if this loop was nested at depth 3 this would match both our parent and our grandparent loop, right? It would pick one arbitrarily?

smowton replied:
Overall this looks a lot like we should be maintaining an active loop stack rather than looking at all currently active loops and their depths.

JohnDumbell replied:
In regards to maintaining an active loop stack: its aim was to do that without an explicit stack we have to modify and keep track of, though at this point it might be worth just doing it, as it's grown more complicated than I intended (and may get more-so). 

Also as you noticed, small issue here that it should be <= in `current_loop_info.nesting_level < loop.second.nesting_level`. since if we've found a equal nesting level we want to keep walking upwards.

Your first read of this is right - it should simply be 'if we've found a loop that is active and greater/at our current depth, we're a child of it, so set it as our parent'.

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,65 @@ void symex_transition(
+    }
+
+    // Otherwise do for-each / while.
+    for(const auto &incoming_edge : state.source.pc->incoming_edges)
```
Check what will happen here for multiple incoming backedges

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,65 @@ void symex_transition(
+
+    // Match the do-while pattern.
+    if(
+      state.source.pc->is_backwards_goto() &&
```
I suspect these cases could be merged into something like

```
while(not_in_loop(to, active_loop_stack.top()))
  active_loop_stack.pop();
```

With our current half-assed loop analysis approximation "not-in-loop" would be something like the location number being greater-than that of the highest-numbered backedge leading to the loop head.

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +67,65 @@ void symex_transition(
       }
+    }
+
+    // And conversely, if we're transitioning out of a loop make sure
```
Note a single GOTO could break out of more than one loop at a time, and might be taken unconditionally -- I don't think that case is handled.

JohnDumbell committed:
loop_analysis.h refactor

Remove redundant P template param.
Move the concept of a loop template that calls back to its base analysis into its own class.

peterschrammel commented on src/goto-checker/bmc_util.h:
```diff
@@ -219,6 +221,15 @@ void run_property_decider(
+  "                              abandons it. Currently uses guard size to\n" \
+  "                              calculate complexity. \n" \
+  " --symex-complexity-failed-children-limit [number]\n" \
+  "                              how many children in a loop / iteration\n" \
```
What are children and parents here?

JohnDumbell replied:
Child branches (in relation to symex states) specifically. Parents is a bit more complicated, as this number is the cumulative failed branches across all active loops up the stack that are allowed. 

So you can have a loop that has 2 child loops (be it directly or in methods it calls), both of these children fail twice, which individually isn't enough to hit any limitation, but when added together their parent loop _does_ breach that, so it gets killed off. 

Also reading this again, the comment is out of date, so I'll update it. Don't want to duplicate too much of the in-line comments though.

peterschrammel commented on src/goto-checker/bmc_util.h:
```diff
@@ -219,6 +221,15 @@ void run_property_decider(
+  "                              how many children in a loop / iteration\n" \
+  "                              are allowed to fail due to complexity\n" \
+  "                              violations before the loop itself and all\n" \
+  "                              parents are abandoned\n" \
```
`abandoned.`

peterschrammel commented on src/goto-symex/complexity_violation.h:
```diff
@@ -0,0 +1,24 @@
+/// Branch: If this particular branch has been deemed too complex.
+enum class complexity_violationt
+{
+  None,
```
See 'enum': https://github.com/diffblue/cbmc/blob/develop/CODING_STANDARD.md

peterschrammel commented on src/goto-symex/goto_symex.h:
```diff
@@ -789,6 +792,8 @@ class goto_symext
   unsigned _total_vccs, _remaining_vccs;
   ///@}
 
+  symex_complexityt complexity_module;
```
Is `complexity_module` better than `symex_complexity`?

peterschrammel commented on src/goto-symex/path_storage.h:
```diff
@@ -115,7 +115,31 @@ class path_storaget
+  }
+
 private:
+  std::unordered_map<irep_idt, std::shared_ptr<lexical_loopst>> loop_analysis;
```
Why `shared_ptr`? Where are the other owners of the object?

JohnDumbell replied:
While `path_storage` is pretty central you absolutely should be able to hold loop information outside of its lifetime, it's not like the data is going to change at all. All I wanted is something that acts like a reference in managed languages, and this fit the bill well enough.

peterschrammel commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,195 @@
+  : log(logger)
+{
+  std::size_t limit = options.get_signed_int_option("symex-complexity-limit");
+  if((complexity_active = limit > 0))
```
weird parentheses

JohnDumbell replied:
Needed for the assignment, doesn't like it otherwise.

peterschrammel commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,195 @@
+#include <cmath>
+
+symex_complexityt::symex_complexityt(
+  message_handlert &logger,
```
logger -> message_handler ?

peterschrammel commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,195 @@
+  {
+    // This gives a curve that allows low limits to be rightly restrictive,
+    // while larger numbers are very large.
+    max_complexity = static_cast<std::size_t>((limit * limit) * 25);
```
Can this overflow?

JohnDumbell replied:
Hard to do on mine, but technically it can. I just assumed people wouldn't give it a number  _that_ high.

peterschrammel commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,195 @@
+  for(auto frame_iter = current_call_stack.rbegin();
+      frame_iter != current_call_stack.rend();
+      ++frame_iter)
+    for(auto &loop_iter : frame_iter->active_loops)
```
I'd appreciate braces here

peterschrammel commented on src/goto-symex/symex_complexity.cpp:
```diff
@@ -0,0 +1,195 @@
+      {
+        sum_complexity += loop_info.children_too_complex;
+        if(sum_complexity > max_loops_complexity)
+          loop_to_blacklist =
```
braces (several occurrences)

peterschrammel commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,20 +85,47 @@ void symex_transition(
     // or
     // 2. we are arriving from an outer loop
+
+    // TODO: This should all be replaced by natural loop analysis.
```
In a future PR?

JohnDumbell replied:
Yeah, full natural loop analysis isn't in a position to use fully yet, why the current implementation is a mash-up of both approaches and also hidden behind the complexity flag.

peterschrammel commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,20 +85,47 @@ void symex_transition(
+      if(
+        incoming_edge->is_backwards_goto() &&
+        incoming_edge->location_number < to->location_number)
+        pop_active_loop(to, frame.active_loops);
```
braces

peterschrammel reviewed (commented):
Are there any tests for this?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ced89d40a68c29ab5742aa25fcba769e4a0ea9c4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/130791262

JohnDumbell committed:
Test for complexity limits

This test will fail if any of the features this relies upon has been
broken, but has the side-affect that if things ARE broken it'll take
a long time to fail. Not much to do about that, as this sort of
long-running symex-destroying code is what this feature is meant for.

JohnDumbell commented:
This should be good for reviews (the failing tests also fail on develop on my machine for whatever reason, so I'm ignoring them for now). No more major changes going in except for review comments.

owen-mc-diffblue commented:
@JohnDumbell All the test failures seem to be expected ones, so I don't immediately see why it isn't passing. The one thing I did see was `symex_complexity/test.desc: Missing EXIT test`, which means that this test.desc file is missing a check for the EXIT variable. The linter also has a complaint: `src/goto-symex/complexity_module.h:12:  #ifndef header guard has wrong style, please use: CPROVER_GOTO_SYMEX_COMPLEXITY_MODULE_H  [build/header_guard] [5]`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d135e434a74eb5e22fafbf16dd6df2d6334d1689).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/131785335

smowton commented on src/analyses/lexical_loops.h:
```diff
@@ -61,9 +61,9 @@ Author: Diffblue Ltd
 ///     * [field] location_number which is an unsigned int.
 template <class P, class T>
-class lexical_loops_templatet : public loop_analysist<P, T>
+class lexical_loops_templatet : public loop_analysist<T>
```
??? `operator()` for example uses template parameter P

JohnDumbell replied:
Yeah but that's only for `lexical_loops_templatet` right? `P` is not used at all in the parent class, so unless template types are scoped really strangely it should be fine?

smowton commented on src/analyses/loop_analysis.h:
```diff
@@ -83,58 +69,105 @@ class loop_templatet
+};
+
+template <typename T>
+class linked_loop_templatet : loop_templatet<T>
```
Linked to what? `loop_with_parent_analysis_templatet`?

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+#include "goto_symex_state.h"
+#include <cmath>
+
+complexity_modulet::complexity_modulet(
```
"module" doesn't really say anything, how about `complexity_limitert`?

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+
+    // In order of priority:
+    //  * explicit limit
+    //  * inferred limit from unwind
```
Does this imply a user can accidentally switch this on with just `--unwind`? If so, why? If not, rephrase.

JohnDumbell replied:
Not sure what the confusion is here? If you have complexity enabled and have an unwind limit, it'll take that unwind limit and infer a best-guess number that can fail. Checking for loop failure boundaries isn't a feature to turn on/off, it's a critical part of how this works, so it needs some value.

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+optionalt<std::reference_wrapper<framet::active_loop_infot>>
+complexity_modulet::get_current_active_loop(call_stackt &current_call_stack)
+{
+  for(auto frame_iter = current_call_stack.rbegin();
```
I think you can use a for-each here?

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+}
+
+/// Takes into account nested loops and various other subtleties.
+optionalt<std::reference_wrapper<framet::active_loop_infot>>
```
Usually optional-of-ref-wrapper is a bit pointless, can just return a non-owning bare pointer instead

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+      // our currently blacklisted ones
+      if(loop_to_blacklist)
+      {
+        loop_info.blacklisted_loops.emplace_back(*loop_to_blacklist);
```
I don't think this works: it adds `loop_to_blacklist`, the originally blacklisted loop, again and again, rather than actually adding parents.

JohnDumbell replied:
Yeah, that's what it's meant to do. Because this is reversed we'll be walking bottom-up, at the point something becomes blacklisted we then add that to every parent that's still executing. Then those parents will make sure they aren't executing any blacklisted loops.

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+
+      // If we've already violated a complexity limit, add to all parent loops
+      // our currently blacklisted ones
+      if(loop_to_blacklist)
```
If this did what the comments say then you wouldn't need to track a particular loop, only a boolean indicating whether we've found trouble.

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+  // are far too complicated to deal with, wasting huge amounts of time.
+  std::size_t sum_complexity = 0;
+
+  // If non-empty, is the lowest-level loop we should blacklist in the stack.
```
Not clear what the comment means. How about `Walk from the innermost to outermost loop on the stack, blacklisting any loop (and all of its parents) that are too complex`?

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+        log.warning()
+          << "[symex-complexity] Loop operations considered too complex"
+          << (state.source.pc->source_location.is_not_nil()
+                ? " at: " + state.source.pc->source_location.as_string()
```
If you can't print a source location at least print instruction->location_number.

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+      }
+    }
+
+    log.warning() << "[symex-complexity] Branch considered too complex"
```
Same here

smowton commented on src/goto-symex/complexity_module.cpp:
```diff
@@ -0,0 +1,205 @@
+    return complexity_violationt::BRANCH;
+  }
+
+  // If we haven't found a bottom-most active loop, just cancel.
```
"just cancel" suggests "cancelling," i.e. truncating a symex path, as used above. In fact here we're *not* cancelling.

smowton commented on src/goto-symex/symex_config.h:
```diff
@@ -48,6 +48,10 @@ struct symex_configt final
 
+  /// \brief Whether this run of symex is under complexity constraints. This
+  /// enables certain analysis' that otherwise aren't run.
+  bool complexty_restraints;
```
sp `complexity`

smowton replied:
`complexity_limits_active`?

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -51,10 +51,28 @@ symex_configt::symex_configt(const optionst &options)
+/// If 'to' is not an instruction in our currently top-most active loop,
+/// pop and re-check until we find an loop we're still active in, or empty
+/// the stack.
+static void pop_active_loop(
```
`pop_exited_loops`

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +85,50 @@ void symex_transition(
+    if(
+      state.source.pc->is_backwards_goto() &&
+      state.source.pc->location_number < to->location_number)
+    {
```
I don't think you need the pattern recognition any more? Can just always call the pop method?

JohnDumbell replied:
Nope, still need it. Lexical loop analysis alone isn't accurate enough, even it does help.

smowton commented on src/goto-symex/complexity_module.h:
```diff
@@ -0,0 +1,117 @@
+
+  /// Functions called when the heuristic has been violated. Can perform
+  /// any form of branch, state or full-program transformation.
+  std::vector<symex_heuristic_transformationt> violation_transformations;
```
`symex_heuristics_transformationt` -> `symex_complexity_limit_exceeded_actiont`. This isn't the heuristic, it's specifically the action to take when a limit (whether heuristic or not) has been broken

smowton reviewed (commented):
Also you should check whether `lexical_loopst` found any control flow not in lexical loop form and act conservatively in that case (most simply: just don't track loops in that function), since there will be loops present it doesn't tell you about

JohnDumbell commented:
Tests added, other comments dealt with. In regards to doing something when the loops aren't in a lexical form - that omission was deliberate. Have a chat to me tomorrow if you want a full breakdown of why/etc. Easier to explain in person.

owen-mc-diffblue commented on src/goto-checker/bmc_util.h:
```diff
@@ -219,6 +221,15 @@ void run_property_decider(
   " --no-self-loops-to-assumptions\n" \
   "                              do not simplify while(1){} to assume(0)\n" \
   " --no-pretty-names            do not simplify identifiers\n" \
+  " --symex-complexity-limit [number]" \
```
`[number]` isn't how the arguments are indicated in the other options, e.g. `--max-field-sensitivity-array-size M`. It is confusing to have it in a different format. It makes it look like it's optional (but I don't think it is). :fish:

owen-mc-diffblue commented on src/goto-checker/bmc_util.h:
```diff
@@ -219,6 +221,15 @@ void run_property_decider(
+  "                              how complex a path can become before symex\n" \
+  "                              abandons it. Currently uses guard size to\n" \
+  "                              calculate complexity. \n" \
+  " --symex-complexity-failed-children-limit [number]\n" \
```
:fish:

owen-mc-diffblue commented on src/goto-symex/symex_config.h:
```diff
@@ -48,6 +48,10 @@ struct symex_configt final
   std::size_t max_field_sensitivity_array_size;
 
+  /// \brief Whether this run of symex is under complexity limits. This
+  /// enables certain analysis' that otherwise aren't run.
```
analysis' -> analyses

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+  // This acts as a context-sensitive loop cancel, so if we've got unwind 20
+  // and find out at the 3rd iteration a particular nested loop is too
+  // complicated, we make sure we don't execute it the other 17 times. But as
+  // soon as we're running the loop again via a context it gets a chance
```
via a different context

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+    max_complexity = static_cast<std::size_t>((limit * limit) * 25);
+
+    const std::size_t failed_children_limit =
+      options.get_signed_int_option("symex-complexity-failed-children-limit");
```
failed-child-loops-limit? "children" by itself suggests e.g. child functions

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+    // while larger numbers are very large.
+    max_complexity = static_cast<std::size_t>((limit * limit) * 25);
+
+    const std::size_t failed_children_limit =
```
child_loops? Also I guess this is an arbitrary complexity number rather than a count? If so use `_complexity` to indicate this isn't a count

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+  }
+}
+
+/// Takes into account nested loops and various other subtleties.
```
Either specify the subtleties or omit

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+
+      // Because we're walking in reverse this will only be non-empty for
+      // parents of the loop that's been blacklisted. We then add that to their
+      // internal lists.
```
...of blacklisted children

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+  if(complexity >= max_complexity)
+  {
+    // If we're too complex, add a counter to the current loop we're in and
+    // check if we've violated children complexity limits.
```
children -> child-loop

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+    return complexity_violationt::BRANCH;
+  }
+
+  // If we haven't found a bottom-most active loop, return with no violation.
```
inner-most (or just, "if we're not in any loop"

smowton commented on src/goto-symex/complexity_limiter.cpp:
```diff
@@ -0,0 +1,214 @@
+  // if so then cancel before we go any further.
+  if(in_blacklisted_loop(current_call_stack, state.source.pc))
+  {
+    log.warning() << "[symex-complexity] Blacklisted loop operation attempted"
```
Trying to enter blacklisted loop

smowton commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+/// As branches get cancelled it keeps track of what loops are currently active
+/// up through the stack. If a loop has too many of its child branches killed
+/// (including additional loops or recursion) it won't do any more unwinds of
+/// that particular loop and blacklist it.
```
and will blacklist it

smowton commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+///         loop C: (complexity: 70)
+///
+/// In the above loop B will be blacklisted if we have a complexity limitation
+/// < 5000, but loop A and B will still be run, because when loop B is removed
```
A and C?

smowton commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+
+/// Symex complexity module.
+///
+/// Dynamically cancels branches when symex considers them too complex.
```
Mention what "cancel" really means (equivalent to ASSUME(false) I'd guess)

smowton commented on src/goto-symex/symex_complexity_limit_exceeded_action.h:
```diff
@@ -0,0 +1,23 @@
+{
+public:
+  virtual void transform(
+    const complexity_violationt heuristic_result,
```
heuristic_result -> complexity_violation

smowton reviewed (approved):
One more round of comment / documentation copyediting; nothing blocking.

JohnDumbell commented:
Going to merge soon as CI passes and I run final tests to make sure with this setting off it has no impact.

owen-mc-diffblue commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+class complexity_limitert
+{
+public:
+  complexity_limitert() = default;
```
Remove this constructor?

JohnDumbell replied:
Couldn't do this in the end

owen-mc-diffblue commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+public:
+  complexity_limitert() = default;
+
+  explicit complexity_limitert(message_handlert &logger) : log(logger)
```
Remove this constructor?

owen-mc-diffblue commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+    goto_symex_statet &current_state);
+
+protected:
+  mutable messaget log;
```
Store as a reference, after removing the default constructor

owen-mc-diffblue commented on src/goto-symex/complexity_limiter.h:
```diff
@@ -0,0 +1,118 @@
+    violation_transformations;
+
+  /// Default heuristic transformation. Adds unconditional false guards to
+  /// state.
```
Update comment: sets state as not reachable

owen-mc-diffblue commented on src/goto-symex/path_storage.h:
```diff
@@ -115,7 +115,37 @@ class path_storaget
+  }
+
 private:
+  std::unordered_map<irep_idt, std::shared_ptr<lexical_loopst>> loop_analysis;
```
Rename to "loop_analyses" or "loop_analysis_map"

owen-mc-diffblue commented on src/goto-symex/path_storage.h:
```diff
@@ -115,7 +115,37 @@ class path_storaget
+  }
+
+  optionalt<std::shared_ptr<lexical_loopst>>
+  try_get_loop_analysis(const irep_idt &function_id)
```
Never used - delete

owen-mc-diffblue commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +85,58 @@ void symex_transition(
+        pop_exited_loops(to, frame.active_loops);
+      }
+
+      // Otherwise do for-each / while.
```
"Match for-each or while"

owen-mc-diffblue commented on src/goto-symex/symex_main.cpp:
```diff
@@ -67,17 +85,58 @@ void symex_transition(
     // or
     // 2. we are arriving from an outer loop
+
+    // TODO: This should all be replaced by natural loop analysis.
```
Mention the current loop analysis is imprecise when saying it should be replaced.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a35c29e6c969f91fe4c8c91694e7ae111632e535).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/132204566
